README.pod
=head1 NAME

B<Router::PathInfo> - PATH_INFO router, based on search trees

=head1 DESCRIPTION

Allows balancing PATH_INFO to static and controllers.
It has a simple and intuitive interface.

=head1 WARNING

Version less than 0.05 is deprecate.

=head1 SYNOPSIS

    use Router::PathInfo;
    
    # or
    use Router::PathInfo as => singletone;
    # this allow to call after new: instance, clear_singleton
    
    my $r = Router::PathInfo->new( 
        static => {
            allready => {
                path => '/path/to/static',
                first_uri_segment => 'static'
            }
        },
        cache_limit => 300
    );
    
    $r->add_rule(
        connect         => '/foo/:enum(bar|baz)/:name(year):re(^\d{4}$)/:any', 
        action          => $some_thing,
        mthods          => ['GET','DELETE'],
        match_callback  => $code_ref
    );
    
    my $env = {PATH_INFO => '/foo/bar/2011/baz', REQUEST_METHOD => 'GET'};
    
    my $res = $r->match($env);
    # or
    my $res = $r->match('/foo/bar/2011/baz'); # GET by default
    
    # $res = {
    #     type => 'controller',
    #     action => $some, # result call $code_ref->($match, $env)
    #     name_segments => {'year' => 2011}
    # }
    
    
    $env = {PATH_INFO => '/static/img/some.jpg'};
    
    $res = $r->match($env);
    
    # $res = {
    #     type  => 'static',
    #     file  => '/path/to/static/img/some.jpg',
    #     mime  => 'image/jpeg'
    # }    

See more details L<Router::PathInfo::Controller>, L<Router::PathInfo::Static>

=head1 PACKAGE VARIABLES

=head2 $Router::PathInfo::as_singleton

Mode as singletone. By default - 0.
You can pick up directly, or:

    use Router::PathInfo as => singletone;
    # or
    require Router::PathInfo;
    Router::PathInfo->import(as => singletone);
    # or
    $Router::PathInfo::as_singleton = 1
    
If you decide to work in singletone mode, raise the flag before the call to C<new>. 

=head1 SINGLETON

When you work in a mode singletone, you have access to methods: C<instance> and C<clear_singleton>

=head1 METHODS

=head2 new(static => $static, cache_limit => $cache_limit)

Constructor. All arguments optsioanlny.

static - it hashref arguments for the constructor L<Router::PathInfo::Static>

cache_limit - limit of matches stored by the rules contain tokens C<:re> and C<:any>, statics and errors. By default - 200.
All matches (that occur on an accurate description) cached without limit.

=head2 add_rule

See C<add_rule> from L<Router::PathInfo::Controller>

=head2 match({PATH_INFO => $path_info, REQUEST_METHOD => $method})

Search match. Initially checked for matches on static, then according to the rules of the controllers.
In any event returns hashref coincidence or an error.

Example:

    {
      type  => 'error',
      code => 400,
      desc  => '$env->{PATH_INFO} not defined'  
    }
    
    {
      type  => 'error',
      code => 404,
      desc  => sprintf('not found for PATH_INFO = %s with REQUEST_METHOD = %s', $env->{PATH_INFO}, $env->{REQUEST_METHOD}) 
    }
    
    {
        type => 'controller',
        action => $action,
        name_segments => $hashref_of_names_segments 
    }
    
    {
        type  => 'static',
        file  => $serch_file,
        mime  => $mime_type
    }

=head1 SOURSE

git@github.com:mrRico/p5-Router-Path-Info.git

=head1 SEE ALSO

L<Router::PathInfo::Static>, L<Router::PathInfo::Controller>

=head1 AUTHOR

mr.Rico <catamoose at yandex.ru>

=cut


Changes
Revision history for Perl extension Router::PathInfo.

0.01  Mon Jun  6 22:11:18 2011
	- original version; created by h2xs 1.23 with options
		-AX Router::PathInfo

0.02  Mon Jun 20 18:09:40 2011
    - add cache
    
0.03  Mon Jun 20 18:09:40 2011
    - fix bug with callback and cache

0.04  Mon Oct 10 12:26:06 2011    
    - Version less than 0.04 is depricated.
    - sub-attribute ':name(...)' has been added
    
0.05  Fri Oct 14 15:18:23 2011    
    - Fix critical bug with cache for controller with callback    
    
0.0502 Thu Dec 13 18:35:18 2012
    - Exclude for a root favicon was added

tools/cacher_benchmark.pl
#!/usr/bin/perl
use strict;
use warnings;

use Test::More;    
use Router::PathInfo;
use Cache::Memcached::Fast;
use Getopt::Long;
use Benchmark qw(:all);

    my $servers;
    GetOptions ("servers=s" => \$servers);
    
    unless ($servers) {
        print "
            call cacher_benchmark.pl like with:
                perl /some/path/cacher_benchmark.pl --servers 10.0.0.15:11211,10.0.0.15:11212
            where 'servers' is a list of your memcache servers

";
        exit;
    }
    $servers = [split ',',$servers];



    my $res;   
    
    my $pi = Router::PathInfo->new();    
    is($pi->add_rule(connect => '/foo/:enum(bar|baz)/:any', action => ['some','bar']), 1, 'check add_rule');
    is($pi->add_rule(connect => '/foo/:enum(bar|baz)/:re(\d{4}\w{4})', action => ['some re','bar re']), 1, 'check add_rule');
    
    $res = $pi->match({PATH_INFO => '/foo/baz/bar', REQUEST_METHOD => 'GET'}); 
    
    # check result
    is(ref $res, 'HASH', 'check ref match');
    is($res->{type}, 'controller', 'check match type');
    is(ref $res->{action}, 'ARRAY', 'check ref action');
    is($res->{action}->[0], 'some', 'check action content 1');
    is($res->{action}->[1], 'bar', 'check action content 2');
    is($res->{segment}->[0], 'baz', 'check segment 1');
    is($res->{segment}->[1], 'bar', 'check segment 2');
    
    $res = $pi->match({PATH_INFO => '/foo/baz/2011year', REQUEST_METHOD => 'GET'}); 
    
    # check result
    is(ref $res, 'HASH', 'check ref match');
    is($res->{type}, 'controller', 'check match type');
    is(ref $res->{action}, 'ARRAY', 'check ref action');
    is($res->{action}->[0], 'some re', 'check action content 1');
    is($res->{action}->[1], 'bar re', 'check action content 2');
    is($res->{segment}->[0], 'baz', 'check segment 1');
    is($res->{segment}->[1], '2011year', 'check segment 2');
    
    my $mem = Cache::Memcached::Fast->new({servers => $servers});
    my $pic = Router::PathInfo->new(
        cacher => $mem
    );    
    is($pic->add_rule(connect => '/foo/:enum(bar|baz)/:any', action => ['some','bar']), 1, 'check add_rule');
    is($pic->add_rule(connect => '/foo/:enum(bar|baz)/:re(\d{4}\w{4})', action => ['some re','bar re']), 1, 'check add_rule');
    
    $res = $pic->match({PATH_INFO => '/foo/baz/bar', REQUEST_METHOD => 'GET'}); 
    
    # check result
    is(ref $res, 'HASH', 'check ref match');
    is($res->{type}, 'controller', 'check match type');
    is(ref $res->{action}, 'ARRAY', 'check ref action');
    is($res->{action}->[0], 'some', 'check action content 1');
    is($res->{action}->[1], 'bar', 'check action content 2');
    is($res->{segment}->[0], 'baz', 'check segment 1');
    is($res->{segment}->[1], 'bar', 'check segment 2');
    
    $res = $pic->match({PATH_INFO => '/foo/baz/2011year', REQUEST_METHOD => 'GET'}); 
    
    # check result
    is(ref $res, 'HASH', 'check ref match');
    is($res->{type}, 'controller', 'check match type');
    is(ref $res->{action}, 'ARRAY', 'check ref action');
    is($res->{action}->[0], 'some re', 'check action content 1');
    is($res->{action}->[1], 'bar re', 'check action content 2');
    is($res->{segment}->[0], 'baz', 'check segment 1');
    is($res->{segment}->[1], '2011year', 'check segment 2');

    pass('*' x 10);
    pass('Start testing');
    pass('*' x 10);
    
    $mem->set('foo', 12, 3600);
    is($mem->get('foo'), 12, 'test memcache read');
    
    my @env = ('/foo/baz/bar','/foo/baz/2011year') x 5;
    cmpthese timethese(
     -1, 
        { 
            WithOutCacher => sub {$pi->match($_) for @env}, 
            WithCacher => sub {$pic->match($_) for @env},
            MemRead => sub {$mem->get('foo') for (1..10)} 
        } 
     );    

    pass('*' x 10);
    print "\n";
    done_testing;


lib/Router/PathInfo/Cacher.pm
package Router::PathInfo::Cacher;

use strict;
use warnings;

use namespace::autoclean;
use Carp;
use Scalar::Util qw(blessed);
use Digest::MD5 qw(md5_hex);

@Router::PathInfo::Cacher::interface = qw(get set namespace);

# если скинуть значение в 0, то кеширования происходить не будет
# для ошибок кеширование настраивается на основе Router::PathInfo::Match::Error->code, т.к.
# 500-ю кешировать нет смысла, а вот 404 - возможно...
$Router::PathInfo::Cacher::ttl_match = {
    'static'        => 10*60,
    'controller'    => 5*60,
    '404'           => 0
};

sub new {
    my $class = shift;
    
    my %param = @_;
    
    # проверка memcache
    unless (blessed($param{cacher})) {
        carp "argument must be cacher instance";
        return;
    }

    for (@Router::PathInfo::Cacher::interface) {
        unless ($param{cacher}->can($_)) {
            carp blessed($param{cacher})." not implemented interface with ".join(',', map {"'$_'"} @Router::PathInfo::Cacher::interface);
            return;
        };
    } 
    
    # создаём объект
    my $self = bless {
        cacher         => $param{cacher},
        ns_prefix      => $param{ns_prefix} || ''
    }, $class;
    
    return $self;
}

sub namespace {$_[0]->{cacher}->namespace($_[0]->{ns_prefix}.':'.$_[1].':')}

=head2 get($uri, $http_method)

Метод, проверяет а базе наличие успешного совпадения для path от uri по соотвествующемц методу

=cut
sub match {
    my ($self, $env) = @_;
    return $self->{cacher}->get($self->_make_key($env));
}

sub _make_key {
    my $self = shift;
    my $env  = shift;
    return md5_hex(join('', $env->{PATH_INFO}, $env->{REQUEST_METHOD}));
}

sub set_match {
    my ($self, $env, $result) = @_;
    
    my $ttl_key = $result->{type} eq 'error' ? $result->{code} : $result->{type}; 
    
    my $ttl = exists $Router::PathInfo::Cacher::ttl_match->{$ttl_key} ? 
                                $Router::PathInfo::Cacher::ttl_match->{$ttl_key} : 
                                $Router::PathInfo::Cacher::ttl_match->{'404'}; 
    
    # если не задано ttl для кеширования - кеширование match'ей не выполняем
    return $ttl ? $self->{cacher}->set($self->_make_key($env), $result, $ttl) : 1;
}

=head1 AUTHOR

mr.Rico <catamoose at yandex.ru>

=cut
1;
__END__


lib/Router/PathInfo/Static.pm
package Router::PathInfo::Static;
use strict;
use warnings;

=head1 NAME

B<Router::PathInfo::Static> - static routing

=head1 DESCRIPTION

Class to describe the routing of statics.
Allows us to describe the statics as follows:

- Specify the starting segment of the URI

- Specify a directory on disk, which will host the search for static

Statics is divided into two parts:

- C<allready> - already exists (the "classic") static

- C<on_demand> - created on demand

Case C<allready> it's different css, js, images. C<on_demand> it's archives and another.
If the file to C<on_demand> not found, C<match> return undef - a signal that makes sense to continue search of L<Router::PathInfo::Controller> routing.

If successful, returns hashref:

    {
        type => 'static',
        mime_type => $mime_type,
        file_name => '/path/to/found.static',
    };
    
This ensures that the file exists, has size, and is readable.


If static is not found (for C<allready>) an error is returned:

    {
        type  => 'error',
        code => 404,
        desc  => sprintf('not found static for PATH_INFO = %s', $env->{PATH_INFO})
    }

If PATH_INFO contains illegal characters (such as C</../> or C</.file>)an error is returned:

    {
        type  => 'error',
        code => 403,
        desc  => sprintf('forbidden for PATH_INFO = %s', $env->{PATH_INFO})   
    }

Return C<undef> means that it makes sense to continue search of L<Router::PathInfo::Controller> routing.  

=head1 SYNOPSIS
    
    my $s = Router::PathInfo::Static->new(
            # describe simple static 
            allready => {
                path => $allready_path,
                first_uri_segment => 'static'
            },
            # describe on demand created static
            on_demand => {
                path => $on_demand_path,
                first_uri_segment => 'archives',
            }
    );
    
    my $env = {PATH_INFO => '/static/some.jpg'};
    my @segment = split '/', $env->{PATH_INFO}, -1; 
    shift @segment;
    $env->{'psgix.tmp.RouterPathInfo'} = {
        segments => [@segment],
        depth => scalar @segment 
    };

    my $res = $s->match($env);
    
    # $res = {
    #     type  => 'static',
    #     file  => $path_to_some_jpg,
    #     mime  => 'image/jpeg'
    # }

=head1 METHODS

=cut

use Plack::MIME;
use File::Spec;
use File::MimeInfo::Magic qw(mimetype);

=head2 new(allready => {path => $dir, first_uri_segment => $uri_segment}, on_demand => {...})

The constructor accepts the description of the statics (allready) and/or static generated on demand (on_demand).
Each description is a hashref with the keys C<'path'> (directory path)
and C<'first_uri_segment'> (the first segment of a PATH_INFO, which defines namespace for designated purposes).    

All arguments are optional. If no arguments are given, the object is not created.

=cut
sub new {
	my $class = shift;
	my %param = @_;
	
	my $hash = {};
	
	for (qw(allready on_demand)) {
		my $token = delete $param{$_};
		if (ref $token) {
	        if (-e $token->{path} and -d _ and $token->{first_uri_segment}) {
	            $hash->{$_.'_path'}        = $token->{path};
	            $hash->{$_.'_uri_segment'} = $token->{first_uri_segment};
	            $hash->{$_}                = 1;
	        } else {
	            $hash->{$_}                = 0;
	        }
		}
	}

	return keys %$hash ? bless($hash, $class) : undef;
}

sub _type_uri {
    my $self          = shift;
    my $first_segment = shift;
    
    for (qw(allready on_demand)) {
    	return $_ if ($self->{$_} and $first_segment eq $self->{$_.'_uri_segment'});
    }
    
    return;
}

=head2 match({'psgix.tmp.RouterPathInfo' => {...}})

Objects method. 
Receives a uri and return:

For C<on_demand> created static, return undef if file not found.

=cut
sub match {
    my $self = shift;
    my $env  = shift;
        
    my @segment = @{$env->{'psgix.tmp.RouterPathInfo'}->{segments}};

    my $serch_file = pop @segment;
    if ($serch_file eq 'favicon.ico' and not @segment) {
        # exclude for root favicon
        $serch_file = File::Spec->catfile($self->{'allready_path'}, $serch_file);
        return -e $serch_file ? {
            type  => 'static',
            file  => $serch_file,
            mime  => Plack::MIME->mime_type($serch_file) || mimetype($serch_file) 
        } : {
            type  => 'error',
            code => 404,
            desc  => sprintf('not found static for PATH_INFO = %s', $env->{PATH_INFO})
        };
    }
    
    return unless ($serch_file and @segment);
    
    # проверим первый сегмент uri на принадлежность к статике
    my $type = $self->_type_uri(shift @segment);
    return unless $type;

    # среди прочего небольшая защита для никсойдов, дабы не отдать секьюрные файлы
    return {
        type  => 'error',
        code => 403,
        desc  => sprintf('forbidden for PATH_INFO = %s', $env->{PATH_INFO})   
    } if ($serch_file =~ /^\./ or $serch_file =~ /~/ or grep {$_ =~ /^\./ or $_ =~ /~/} @segment);

    $serch_file = File::Spec->catfile($self->{$type.'_path'}, @segment, $serch_file);
    if (-f $serch_file and -s _ and -r _) {
        return {
            type  => 'static',
            file  => $serch_file,
            mime  => Plack::MIME->mime_type($serch_file) || mimetype($serch_file)
        }
    } else {
        return $type eq 'allready' ?
            {
                type  => 'error',
                code => 404,
                desc  => sprintf('not found static for PATH_INFO = %s', $env->{PATH_INFO})
            } : 
            undef;
    }
}

=head1 DEPENDENCIES

L<Plack::MIME>, L<File::MimeInfo::Magic>

=head1 SEE ALSO

L<Router::PathInfo>, L<Router::PathInfo::Controller>

=head1 AUTHOR

mr.Rico <catamoose at yandex.ru>

=cut
1;
__END__


tools/controller_benchmark.pl
#!/usr/bin/perl
use strict;
use warnings;

use Test::More;

    pass('*' x 10);
    
use Router::PathInfo::Controller;
use Router::Simple;
use Benchmark qw(:all);
my $res;

    my $router = Router::Simple->new();
    $router->connect('/foo/bar/:int', {controller => '1', action => '11'});
    $router->connect('/foo/baz/:sd', {controller => '2', action => '22'});    
    
    my $r = Router::PathInfo::Controller->new();
    $r->add_rule(connect => '/foo/bar/:any', action => ['1','11']);    
    $r->add_rule(connect => '/foo/baz/:any', action => ['2','2']);
    
    my @env = map { {PATH_INFO => $_, REQUEST_METHOD => 'GET'} } ('/foo/bar/200', '/foo/baz/400') x 4;

	for (@env) {
	    my @segment = split '/', $_->{PATH_INFO}, -1; 
	    shift @segment;
	    $_->{'psgix.tmp.RouterPathInfo'} = {
	        segments => [@segment],
	        depth => scalar @segment 
	    };
	}
    
    pass('test resolve exactly match');
    
    $res = $r->match($env[0]);
    is($res->{action}->[0], '1', 'check Router::PathInfo::Controller');
    $res = $router->match($env[0]);
    is($res->{controller}, '1', 'check Router::Simple');
    pass('*' x 10);
    
    pass('Benchmark exactly match');
    cmpthese timethese(
     -1, 
        { 
            My => sub {$r->match($_) for @env}, 
            Other => sub {$router->match($_) for @env} 
        } 
     );
    pass('*' x 10);
    
    my @env404 = map { {PATH_INFO => $_, REQUEST_METHOD => 'GET'} } ('/','/foo/main/moi/red/grep') x 4;

    for (@env404) {
        my @segment = split '/', $_->{PATH_INFO}, -1; 
        shift @segment;
        $_->{'psgix.tmp.RouterPathInfo'} = {
            segments => [@segment],
            depth => scalar @segment 
        };
    }     
      
    pass('test resolve 404');
    
    $res = $r->match($env404[0]);
    is($res, undef, 'check Router::PathInfo::Controller');
    $res = $router->match($env404[0]);
    is($res, undef, 'check Router::Simple');
    pass('*' x 10);
    
    pass('Benchmark only 404 match');
    cmpthese timethese(
     -1, 
        { 
            My => sub {$r->match($_) for @env404}, 
            Other => sub {$router->match($_) for @env404} 
        } 
     );
     pass('*' x 10);
     
     pass('added more rules');
     $r->add_rule(connect => '/foo/bar/fun/more/rules', action => ['3','33']);
     $r->add_rule(connect => '/foo/bar/fun/more/rules/', action => ['4','44']);
     $r->add_rule(connect => '/foo/bar/fun/more/rules/for/test', action => ['5','55']);
     
     $router->connect('/foo/bar/fun/more/rules', {controller => '3', action => '33'});
     $router->connect('/foo/bar/fun/more/rules/', {controller => '4', action => '44'});
     $router->connect('/foo/bar/fun/more/rules/for/test', {controller => '5', action => '55'});

    @env = map { {PATH_INFO => $_, REQUEST_METHOD => 'GET'} } ('/foo/bar/200', '/foo/baz/400', '/', '/foo/bar/fun/more/rules/for/test', '/foo/bar/fun/more/rules/');

    for (@env) {
        my @segment = split '/', $_->{PATH_INFO}, -1; 
        shift @segment;
        $_->{'psgix.tmp.RouterPathInfo'} = {
            segments => [@segment],
            depth => scalar @segment 
        };
    }     
     
     pass('Benchmark more rules');
     
    cmpthese timethese(
     -1, 
        { 
            My => sub {$r->match($_) for @env404}, 
            Other => sub {$router->match($_) for @env404} 
        } 
     );     
     
    pass('*' x 10);
    print "\n";
    done_testing;


tools/withstatic_benchmark.pl
#!/usr/bin/perl
use strict;
use warnings;

use Test::More;

    pass('*' x 10);
    
use Router::PathInfo;
use Router::Simple;
use Benchmark qw(:all);
my $res;

    my $router = Router::Simple->new();
    $router->connect('/foo/bar/:int', {controller => '1', action => '11'});
    $router->connect('/foo/baz/:sd', {controller => '2', action => '22'});    
    
    my $r = Router::PathInfo->new();
    $r->add_rule(connect => '/foo/bar/:any', action => ['1','11']);    
    $r->add_rule(connect => '/foo/baz/:any', action => ['2','2']);
    
    my @env = map { {PATH_INFO => $_, REQUEST_METHOD => 'GET'} } ('/foo/bar/200', '/foo/baz/400') x 4;

    pass('test resolve exactly match');
    
    $res = $r->match($env[0]);
    is($res->{action}->[0], '1', 'check Router::PathInfo');
    $res = $router->match($env[0]);
    is($res->{controller}, '1', 'check Router::Simple');
    pass('*' x 10);
    
    pass('Benchmark exactly match');
    cmpthese timethese(
     -1, 
        { 
            'Router::PathInfo' => sub {$r->match($_) for @env}, 
            'Router::Simple' => sub {$router->match($_) for @env} 
        } 
     );
    pass('*' x 10);
    
    my @env404 = map { {PATH_INFO => $_, REQUEST_METHOD => 'GET'} } ('/','/foo/main/moi/red/grep') x 4;

    pass('test resolve 404');
    
    $res = $r->match($env404[0]);
    is($res, undef, 'check Router::PathInfo');
    $res = $router->match($env404[0]);
    is($res, undef, 'check Router::Simple');
    pass('*' x 10);
    
    pass('Benchmark only 404 match');
    cmpthese timethese(
     -1, 
        { 
            'Router::PathInfo' => sub {$r->match($_) for @env404}, 
            'Router::Simple' => sub {$router->match($_) for @env404} 
        } 
     );
     pass('*' x 10);
     
     pass('added more rules');
     $r->add_rule(connect => '/foo/bar/fun/more/rules', action => ['3','33']);
     $r->add_rule(connect => '/foo/bar/fun/more/rules/', action => ['4','44']);
     $r->add_rule(connect => '/foo/bar/fun/more/rules/for/test', action => ['5','55']);
     
     $router->connect('/foo/bar/fun/more/rules', {controller => '3', action => '33'});
     $router->connect('/foo/bar/fun/more/rules/', {controller => '4', action => '44'});
     $router->connect('/foo/bar/fun/more/rules/for/test', {controller => '5', action => '55'});

    @env = map { {PATH_INFO => $_, REQUEST_METHOD => 'GET'} } ('/foo/bar/200', '/foo/baz/400', '/', '/foo/bar/fun/more/rules/for/test', '/foo/bar/fun/more/rules/');

     pass('Benchmark more rules');
     
    cmpthese timethese(
     -1, 
        { 
            'Router::PathInfo' => sub {$r->match($_) for @env404}, 
            'Router::Simple' => sub {$router->match($_) for @env404} 
        } 
     );     
     
    pass('*' x 10);
    print "\n";
    done_testing;


lib/Router/PathInfo.pm
package Router::PathInfo;
use strict;
use warnings;

our $VERSION = '0.0502';

use Carp;

use Router::PathInfo::Controller;
use Router::PathInfo::Static;

=head1 NAME

B<Router::PathInfo> - PATH_INFO router, based on search trees

=head1 DESCRIPTION

Allows balancing PATH_INFO to static and controllers.
It has a simple and intuitive interface.

=head1 WARNING

Version less than 0.05 is deprecate.

=head1 SYNOPSIS

    use Router::PathInfo;
    
    # or
    use Router::PathInfo as => singletone;
    # this allow to call after new: instance, clear_singleton
    
    my $r = Router::PathInfo->new( 
        static => {
            allready => {
                path => '/path/to/static',
                first_uri_segment => 'static'
            }
        },
        cache_limit => 300
    );
    
    $r->add_rule(
        connect         => '/foo/:enum(bar|baz)/:name(year):re(^\d{4}$)/:any', 
        action          => $some_thing,
        mthods          => ['GET','DELETE'],
        match_callback  => $code_ref
    );
    
    my $env = {PATH_INFO => '/foo/bar/2011/baz', REQUEST_METHOD => 'GET'};
    
    my $res = $r->match($env);
    # or
    my $res = $r->match('/foo/bar/2011/baz'); # GET by default
    
    # $res = {
    #     type => 'controller',
    #     action => $some, # result call $code_ref->($match, $env)
    #     name_segments => {'year' => 2011}
    # }
    
    
    $env = {PATH_INFO => '/static/img/some.jpg'};
    
    $res = $r->match($env);
    
    # $res = {
    #     type  => 'static',
    #     file  => '/path/to/static/img/some.jpg',
    #     mime  => 'image/jpeg'
    # }    

See more details L<Router::PathInfo::Controller>, L<Router::PathInfo::Static>

=head1 PACKAGE VARIABLES

=head2 $Router::PathInfo::as_singleton

Mode as singletone. By default - 0.
You can pick up directly, or:

    use Router::PathInfo as => singletone;
    # or
    require Router::PathInfo;
    Router::PathInfo->import(as => singletone);
    # or
    $Router::PathInfo::as_singleton = 1
    
If you decide to work in singletone mode, raise the flag before the call to C<new>. 

=cut

my $as_singletone = 0;

sub import {
    my ($class, %param) = @_;
    $as_singletone = 1 if ($param{as} and $param{as} eq 'singletone');
    return;
}

=head1 SINGLETON

When you work in a mode singletone, you have access to methods: C<instance> and C<clear_singleton>

=cut


=head1 METHODS

=head2 new(static => $static, cache_limit => $cache_limit)

Constructor. All arguments optsioanlny.

static - it hashref arguments for the constructor L<Router::PathInfo::Static>

cache_limit - limit of matches stored by the rules contain tokens C<:re> and C<:any>, statics and errors. By default - 200.
All matches (that occur on an accurate description) cached without limit.

=cut

my $singleton = undef;

sub new {
    return $singleton if ($as_singletone and $singleton);
    
    my $class = shift;
    my $param = {@_};
    
    my $self = bless {
        static      => UNIVERSAL::isa($param->{static}, 'HASH')     ? Router::PathInfo::Static->new(%{delete $param->{static}}) : undef,
        controller  => UNIVERSAL::isa($param->{controller}, 'HASH') ? Router::PathInfo::Controller->new(%{delete $param->{controller}}) : Router::PathInfo::Controller->new(),
        cache                   => {},
        _hidden_cache           => {},
        cache_limit             => (defined $param->{cache_limit} and $param->{cache_limit}) =~ /^\d+$/ ? $param->{cache_limit} : 200,
        cache_cnt               => 0
    }, $class;
    
    $singleton = $self if $as_singletone;
     
    return $self;
}

=head2 add_rule

See C<add_rule> from L<Router::PathInfo::Controller>

=cut
sub add_rule {
    my $self = shift;
    my $ret = 0;
    if ($self->{controller}) {
        $self->{cache_cnt}  = 0;
        $self->{cache}      = {};
        $self->{controller}->add_rule(@_);
    } else {
        carp "controller not defined";
    }
}

sub instance        {$as_singletone ? $singleton : carp "singletone not allowed"}
sub clear_singleton {undef $singleton}

=head2 match({PATH_INFO => $path_info, REQUEST_METHOD => $method})

Search match. Initially checked for matches on static, then according to the rules of the controllers.
In any event returns hashref coincidence or an error.

Example:

    {
      type  => 'error',
      code => 400,
      desc  => '$env->{PATH_INFO} not defined'  
    }
    
    {
      type  => 'error',
      code => 404,
      desc  => sprintf('not found for PATH_INFO = %s with REQUEST_METHOD = %s', $env->{PATH_INFO}, $env->{REQUEST_METHOD}) 
    }
    
    {
        type => 'controller',
        action => $action,
        name_segments => $hashref_of_names_segments 
    }
    
    {
        type  => 'static',
        file  => $serch_file,
        mime  => $mime_type
    }

=cut
sub match {
    my $self = shift; 
    my $env  = shift;
    
    unless (ref $env) {
        $env = {PATH_INFO => $env, REQUEST_METHOD => 'GET'};
    } else {
        $env->{REQUEST_METHOD} ||= 'GET';
    }
    
    my $match = undef;
    
    $match = {
      type  => 'error',
      code => 400,
      desc  => '$env->{PATH_INFO} not defined'  
    } unless $env->{PATH_INFO};
    
    # find in cache
    my $cache_key = join('#',$env->{PATH_INFO}, $env->{REQUEST_METHOD});
    my $cache_match = $self->{cache}->{$cache_key} || $self->{_hidden_cache}->{$cache_key};
    if ($cache_match) {
        # only for controller
        $cache_match = $cache_match->{_callback}->({%$cache_match},$env) if exists $cache_match->{_callback};
        return $cache_match;
    };
    
    my @segment = split '/', $env->{PATH_INFO}, -1; shift @segment;
    $env->{'psgix.tmp.RouterPathInfo'} = {
        segments => [@segment],
        depth => scalar @segment 
    };
    
    # check in static
    if (not $match and $self->{static}) {
        $match = $self->{static}->match($env);
    }
    
    # check in controllers
    # $not_exactly - match with regexp
    my $not_exactly = 0;
    if (not $match and $self->{controller}) {
        ($not_exactly, $match) = $self->{controller}->match($env);
    }
    
    # not found?
    $match ||= {
      type  => 'error',
      code => 404,
      desc  => sprintf('not found for PATH_INFO = %s with REQUEST_METHOD = %s', $env->{PATH_INFO}, $env->{REQUEST_METHOD}) 
    };
    
    delete $env->{'psgix.tmp.RouterPathInfo'};
    
    # cache!
    if (not $not_exactly and $match->{type} eq 'controller') {
        $self->{_hidden_cache}->{$cache_key} = $match;
    } elsif ($self->{cache_limit}) {
        if ($self->{cache_cnt} > $self->{cache_limit}) {
            $self->{cache_cnt} = 0;
            $self->{cache} = {};
        } else {
            $self->{cache_cnt}++;
        }
        $self->{cache}->{$cache_key} = $match;        
    }
    
    # only for controller    
    $match = $match->{_callback}->({%$match},$env) if exists $match->{_callback};
    
    # match is done
    return $match;
}

=head1 SOURSE

git@github.com:mrRico/p5-Router-Path-Info.git

=head1 SEE ALSO

L<Router::PathInfo::Static>, L<Router::PathInfo::Controller>

=head1 AUTHOR

mr.Rico <catamoose at yandex.ru>

=cut
1;
__END__
